### PR TITLE
[fix][txn] Implement compatibility for transaction buffer segmented snapshot feature upgrade

### PR DESCRIPTION
**Pull request by liangyepianzhou**
master https://github.com/apache/pulsar/issues/16913
## Motivation:
The transaction buffer segmented snapshot feature aims to improve the transaction buffer's performance by segmenting the snapshot and managing it more efficiently. However, for existing topics that were created before this feature was introduced, we need to ensure a seamless transition and compatibility when enabling the segmented snapshot feature.

## Modifications:
1. Updated the `recoverFromSnapshot()` method to read from another topic if the `persistentSnapshotIndexes` is null. This ensures that the appropriate snapshot data is fetched during the recovery process when upgrading to the segmented snapshot feature.
2. Created a new test `testSnapshotProcessorUpdate()` that verifies the compatibility of the transaction

### Verifying this change

- [ ] Make sure that the change passes the CI checks.

*(Please pick either of the following options)*

This change is a trivial rework / code cleanup without any test coverage.

*(or)*

This change is already covered by existing tests, such as *(please describe tests)*.

*(or)*

This change added tests and can be verified as follows:

*(example:)*
  - *Added integration tests for end-to-end deployment with large payloads (10MB)*
  - *Extended integration test for recovery after broker failure*

### Does this pull request potentially affect one of the following parts:

*If the box was checked, please highlight the changes*

- [ ] Dependencies (add or upgrade a dependency)
- [ ] The public API
- [ ] The schema
- [ ] The default values of configurations
- [ ] The threading model
- [ ] The binary protocol
- [ ] The REST endpoints
- [ ] The admin CLI options
- [ ] The metrics
- [ ] Anything that affects deployment

### Documentation

- [ ] `doc` 
- [ ] `doc-required` 
- [x] `doc-not-needed` 
- [ ] `doc-complete` 

### Matching PR in forked repository

PR in forked repository:

### CODE REVIEW

File: pulsar-broker/src/test/java/org/apache/pulsar/broker/transaction/SegmentAbortedTxnProcessorTest.java [hunk 1]
```diff
@@ -286,4 +292,79 @@ private void doCompaction(TopicName topic) throws Exception {
...
+     * 1. Creates a topic with segmented snapshot disabled.
+     * 2. Sends 10 messages without using transactions.
+     * 3. Sends 10 messages using transactions and aborts them.
+     * 4. Verifies that only the non-transactional messages are received.
+     * 5. Enables the segmented snapshot feature and sets the snapshot segment size.
+     * 6. Unloads the topic and re-verifies that only the non-transactional messages are received.
+     * 7. Sends a new message, and checks if the topic has exactly one segment.
+     */
+    @Test
+    public void testSnapshotProcessorUpdate() throws Exception {
```
Comment (congbobo184) [inline pulsar-broker/src/test/java/org/apache/pulsar/broker/transaction/SegmentAbortedTxnProcessorTest.java:309]:
```suggestion
    public void testSnapshotProcessorUpgrade() throws Exception {
```

File: pulsar-broker/src/test/java/org/apache/pulsar/broker/transaction/SegmentAbortedTxnProcessorTest.java [hunk 2]
```diff
@@ -286,4 +292,79 @@ private void doCompaction(TopicName topic) throws Exception {
...
+                    .build().get();
+            producer.newMessage(txn).value(("test-txn-message-" + i).getBytes()).sendAsync();
+            txn.abort().get();
+        }
+
+        // Verify the data
+        for (int i = 0; i < 10; i++) {
+            Message<byte[]> msg = consumer.receive(5, TimeUnit.SECONDS);
+            assertEquals("test-message-" + i, new String(msg.getData()));
+            consumer.acknowledge(msg);
```
Comment (congbobo184) [inline pulsar-broker/src/test/java/org/apache/pulsar/broker/transaction/SegmentAbortedTxnProcessorTest.java:337]: why ack it?

File: pulsar-broker/src/test/java/org/apache/pulsar/broker/transaction/SegmentAbortedTxnProcessorTest.java [hunk 3]
```diff
@@ -286,4 +292,79 @@ private void doCompaction(TopicName topic) throws Exception {
...
+
+        // Send 10 messages using transactions and abort them
+        for (int i = 0; i < 10; i++) {
+            Transaction txn = pulsarClient.newTransaction()
+                    .withTransactionTimeout(5, TimeUnit.SECONDS)
+                    .build().get();
+            producer.newMessage(txn).value(("test-txn-message-" + i).getBytes()).sendAsync();
+            txn.abort().get();
+        }
+
```
Comment (congbobo184) [inline pulsar-broker/src/test/java/org/apache/pulsar/broker/transaction/SegmentAbortedTxnProcessorTest.java:332]: send 10 - 20, verify can receive 0 - 20

File: pulsar-broker/src/test/java/org/apache/pulsar/broker/transaction/SegmentAbortedTxnProcessorTest.java [hunk 4]
```diff
@@ -286,4 +292,79 @@ private void doCompaction(TopicName topic) throws Exception {
...
+        // Enable segmented snapshot
+        this.pulsarService.getConfig().setTransactionBufferSegmentedSnapshotEnabled(true);
+        this.pulsarService.getConfig().setTransactionBufferSnapshotSegmentSize(8 + PROCESSOR_TOPIC.length() +
+                SEGMENT_SIZE * 3);
+
+        // Unload the topic and re-verify the data
+        admin.topics().unload(topicName);
+        consumer.close();
+        consumer = pulsarClient.newConsumer().topic(topicName).subscriptionName("test-sub").subscribe();
+        for (int i = 0; i < 10; i++) {
```
Comment (congbobo184) [inline pulsar-broker/src/test/java/org/apache/pulsar/broker/transaction/SegmentAbortedTxnProcessorTest.java:349]: verify can receive 0 - 20

File: pulsar-broker/src/test/java/org/apache/pulsar/broker/transaction/SegmentAbortedTxnProcessorTest.java [hunk 5]
```diff
@@ -286,4 +292,79 @@ private void doCompaction(TopicName topic) throws Exception {
...
+        txn.abort().get();
+
+        // Check if the topic has only one segment
+        Awaitility.await().untilAsserted(() -> {
+            String segmentTopic = "persistent://" + NAMESPACE1 + "/" +
+                    SystemTopicNames.TRANSACTION_BUFFER_SNAPSHOT_SEGMENTS;
+            TopicStats topicStats = admin.topics().getStats(segmentTopic);
+            assertEquals(1, topicStats.getMsgInCounter());
+        });
+    }
```
Comment (congbobo184) [inline pulsar-broker/src/test/java/org/apache/pulsar/broker/transaction/SegmentAbortedTxnProcessorTest.java:369]: check this before receive

File: pulsar-broker/src/test/java/org/apache/pulsar/broker/transaction/SegmentAbortedTxnProcessorTest.java (hunk 3 above)
Comment (liangyepianzhou) [reply pulsar-broker/src/test/java/org/apache/pulsar/broker/transaction/SegmentAbortedTxnProcessorTest.java:332]: Sorry, I doesn’t understand this.

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/transaction/buffer/impl/SnapshotSegmentAbortedTxnProcessorImpl.java [hunk 6]
```diff
@@ -265,7 +266,7 @@ public CompletableFuture<PositionImpl> recoverFromSnapshot() {
                     PositionImpl finalStartReadCursorPosition = startReadCursorPosition;
                     TransactionBufferSnapshotIndexes finalPersistentSnapshotIndexes = persistentSnapshotIndexes;
                     if (persistentSnapshotIndexes == null) {
```
Comment (poorbarcode) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/transaction/buffer/impl/SnapshotSegmentAbortedTxnProcessorImpl.java:268]: After turning on the feature `transactionBufferSegmentedSnapshotEnabled,` if users create a new topic, the method `recoverOldSnapshot` will trigger the creation of the topic `__transaction_buffer_snapshot`, but it is not needed, right?

And can we add a test to guarantee this issue will not occur?

**Review by poorbarcode — APPROVED** [review]

**Review by congbobo184 — APPROVED** [review]

**Pull request merged**